### PR TITLE
LibPDF: Cache rendered image bitmaps

### PR DESCRIPTION
**Pull request by IdanHo**
This should help quite a bit with #23157.

### CODE REVIEW

File: Userland/Libraries/LibPDF/ObjectDerivatives.h [hunk 1]
```diff
@@ -207,6 +208,27 @@ class IndirectValue final : public Object {
     Value m_value;
 };
 
+class BitmapObject final : public Object {
```
Comment (IdanHo) [inline Userland/Libraries/LibPDF/ObjectDerivatives.h:211]: I don't love having to add a new type of Object, but I couldn't think of a much nicer way to do this.

**Comment by nico** [comment]
Can you put the first two commits in a separate PR from the last two?

The last two are great / embarrassing and should go in as fast as possible :)

The first two are likely also very effective. But we currently cache nothing (*) across page renders, which is also kind of a nice property – it means memory use doesn't go up as you go through many pages of a big document. And most of the time, images aren't shared across pages. (One image is sometimes used several times on a single page and we don't cache that either; maybe we could do that.)

I kind of have "add a bounded cache to LibPDF" on my list somewhere, but imho that should be shared across docs, have a cap on size, etc.

*: The exception to this are uncompressed stream data, which I'm also not 100% happy about.

**Comment by IdanHo** [comment]
@nico I've split them up, I agree this specific cache impl might not be optimal. It's just a bit sad to see us execute the exact same bitmap rendering routines with the exact same inputs 500+ times in 0000364.pdf

**Comment by nico** [comment]
On the approach:

A file I like (or dislike) for testing CLUT perf is 0000711.pdf. It has many full-page CMYK images, which are currently unusably slow to convert (takes 3s for some of the pages to render on my very fast system). (For matrix profiles, I like/dislike  0000849.pdf – matrix profile perf is much better than it used to be.) Another CLUT stress PDF is https://shifthappens.site/gorton-perfected-specimen.pdf which also has many full-page images, but they're RGB instead of CMYK (but still use a CLUT color profile somehow).

This patch does nothing to help the "full-page bitmap with a CLUT profile" use case. We will have to make this code an order of magnitude (or ideally, 20-30x) faster somehow.

This patch helps a _lot_ with rendering all pages of `0000364.pdf`. `time Build/lagom/bin/pdf --debugging-stats /Users/thakis/Downloads/0000/0000364.pdf` goes from several minutes to 2.4s on my system. Thanks to that, `time Meta/test_pdf.py ~/Downloads/0000 ` does go from 4m22s to 1m28s. That's a bit speedup.

It mostly only helps that one file though. If I exclude it from Meta/test_pdf.py as described in #23157, then this patch here takes test_pdf.py time from 1m23s to 1m17s. That's a real improvement, but a small one.

The usecase of "same big image used on all pages" is as far as I can tell rare in practice. (I thought this would also help with page 1 of `0000826.pdf` which I think draws the same large bitmap on top of itself a bunch of times – with blending, but we don't implement the blending part yet.)

Given that we have to make the CLUT code dramatically faster anyways: once that's done `0000354.pdf` should only take 10-20s to render (more than the 2.4s this patch gives us, but plenty fast enough for a user to page though it), I'm not sure this patch is solving a real problem. (The problem of "test_pdf.py takes too long to run" could be solved by filtering out that one pdf too – and waiting 4m22s before sending out a PR is pretty annoying but also doable. But maybe we should exclude it from the .py script.)

(As mentioned before, the memory use argument is a bit weak at the moment since streams do store decompressed data forever once a stream has been decompressed once. But I think that's something we should change instead, at least for objects referenced only from one page.)

So I'm not sure this patch here is worth doing at all.

On the implementation:

If we do want to do it, we shouldn't do it by adding a made-up name to an object. If we ever want to do a serializer, we'll have to remember to filter these out, and it's generally a bad time. Another approach that seems way nicer could be to have a stream object -> bitmap hashmap on Document?

**Comment by nico** [comment]
#23253 makes things twice as fast in many cases, and speeds up 0000364.pdf enough that it's just barely the long pole. So I feel even more that this patch here isn't worth it now :)

**Comment by IdanHo** [comment]
Yep, just got back to this, #23253 is just better, so no need for this anymore.

**Pull request closed without merging**